### ISSUE TITLE
Wedges with ends linked to different voices

### ISSUE THREAD

**Issue by hbitteur** [Audiveris/audiveris#151]
Using the same score as in #118 we can observe that some wedges (either crescendo or diminuendo) are not correctly rendered by music editors.
Though these wedges are present in the exported MusicXML file, MuseScore spans them through many measures, while Finale does not render them at all.

Here is an example (using input view and output view):
![image](https://user-images.githubusercontent.com/22053634/38983324-5afc857c-43c4-11e8-9946-b88e4bd3788a.png)
![image](https://user-images.githubusercontent.com/22053634/38983407-89812c18-43c4-11e8-8326-cb0584293904.png)

One can notice that the wedge is linked on left and right side to head chords. But these two head chords do not belong to the same voice: voice 2 for the left side and voice 1 for the right side.
In a given measure, MusicXML processes voices one after the other, which results in:

```xml
      <direction>
        <direction-type>
          <wedge type="stop" spread="0" default-x="39"></wedge><!-- stop -->
        </direction-type>
        <staff>1</staff>
      </direction>
      <note default-x="139">
        <pitch>
          <step>G</step>
          <octave>4</octave>
        </pitch>
        <duration>2</duration>
        <voice>1</voice>
        <type>quarter</type>
        <stem default-y="5">up</stem>
        <staff>1</staff>
        <notations>
          <slur type="stop" number="1" bezier-x="-5" bezier-y="13" default-x="7" default-y="-5"></slur>
        </notations>
      </note>
      <backup>
        <duration>8</duration>
      </backup>
      <forward>
        <duration>4</duration>
        <voice>2</voice>
        <staff>1</staff>
      </forward>
      <direction placement="below">
        <direction-type>
          <wedge type="diminuendo" spread="12" default-x="7" default-y="-84"></wedge> <!-- start-->
        </direction-type>
        <staff>1</staff>
      </direction>
      <note default-x="97">
        <pitch>
          <step>B</step>
          <alter>-1</alter>
          <octave>3</octave>
        </pitch>
        <duration>4</duration>
        <voice>2</voice>
        <type>half</type>
        <accidental>flat</accidental>
        <stem default-y="-84">down</stem>
        <staff>1</staff>
      </note>
```
In short, in the MusicXML stream, the "stop" event occurs **before** the "start" event, although their location abscissae are correct.
How could we fix this? Any idea?

**Comment by JalonSolov** [comment]
In this particular case, if stop < start, is seems like you could simply swap them.  If necessary, an extra pass before output to do the swaps.

**Comment by JalonSolov** [comment]
Alternatively... why is the crescendo symbol (often call a 'hairpin' in English) being attached to a specific voice?  It applies to all voices.  Either attach it to a measure, or if it must be attached to a voice, always attach it to the lowest numbered voice where the hairpin begins/ends.  In the case shown, it would be attached to voice 1 at both ends, and there wouldn't be a problem.

**Comment by hbitteur** [comment]
The wedge "attachment" to chords is governed by MusicXML, please refer to http://usermanuals.musicxml.com/MusicXML/MusicXML.htm#EL-MusicXML.htm%3FTocPath%3DMusicXML%2520Reference%7CScore%2520Schema%2520(XSD)%7CElements%7C_____0

We cannot change this logic on our own.

My understanding is that, for various reasons we discussed recently, I have to improve the way voices are defined. It's not an easy task.
See the Dichterliebe01 example (in Audiveris examples), to get an idea of the voices complexity. Its MusicXML content is available on https://github.com/dn-m/MusicXML/blob/master/Dichterliebe01.xml

**Comment by hbitteur** [comment]
You said "_the hairpin neither begins nor ends on a particular chord/note_". The question is "how can we tell if the hairpin begins or ends on a particular note?". To my knowledge there is no well recognized abscissa margin which could drive the decision.

I have some problem with your quote of MusicXML direction: "_the direction element can be associated
with the note element that follows it in score order that is not in a different voice_". How should we understand "_the note element \[...\] not in different voice_"? Compared with which voice? The voice of the direction???

Anyway, many thanks for the pointers you provided.

Current (simplistic) strategy in Audiveris is to look at staff above (if any) and pick up the closest chord (abscissawise). Otherwise, look at staff below and pick up the closest chord.
At this time (CURVES step), no voice is considered, because voices will be discovered in later step (RHYTHMS). But we could postpone wedge attachment until RHYTHMS step when these voices are finally known, this would be an improvement.

Back to our concrete example, what should we do? :
![image](https://user-images.githubusercontent.com/22053634/38983407-89812c18-43c4-11e8-8326-cb0584293904.png)
We could always stick to voice 1 as @Bacchushlg suggested (this would mean the chords in blue)
In this particular case, the left side of the wedge is within the abscissa width of chords above and below the wedge, so any of these chords would equally fit (all other things being equal).
But for the right side, the wedge goes far beyond the last chord above, and is in fact very near the last chord below. So I would be tempted to "attach" the wedge to the two chords below (which belong to the same voice, in maroon color).

**Comment by JalonSolov** [comment]
In the "Mary had a little lamb" example, the "mp" actually starts the crescendo, and the "ff" ends it. Having the hairpin in between just says to make it a smooth transition between the two, rather than an abrupt jump.

I think what the text means in the MusicXML DTDs is that the hairpin should just be tied to positions within the measure, rather than tied to specific chords.  However, you _can_ tie it to specific notes, and the way to do that is to put the <direction> element immediately after the <note> element in the output (as in your example XML).  Otherwise, it could show up anywhere within the <measure> element, or within the outer <part> element, depending on whether it spans measures.

That would have to be tested, I think, to see how MusicXML readers handle it.  It makes sense from what I have read, but not all readers may have been written that way.

One place to have this sort of discussion with more knowledgeable people is on [NOTATIO](http://notat.io/) "A Forum devoted to the Practice of Music Notation".  They  have an entire board on "Digital Notation Tools".

For what audiveris does, where it really wants to tie it to notes, it would be best to use your last thought, and tie it to the maroon color notes.  Then it satisfies the "note element that follows it in score order that is not in a different voice" requirement.  You just have to adjust the 'default-y' values to put it in the right place.
